lookout/style/format/benchmarks/top_repos_quality.py

```diff
@@ -0,0 +1,349 @@
+        if pattern in line:
+            line = line.replace("`", "")
+            res = list(map(float, line.split(pattern)))
+            assert len(res) == 2
+            n_rules, avg_len = res
+            n_rules = int(n_rules)
+    return n_rules, avg_len
+
+
+def main(args):
+    """Entry point for quality report generation."""
+    slogging.setup(args.log_level, False)
+    if not server.exefile.exists():
+        server.fetch()  # download executable
+    # prepare output directory
+    dir_to_save = args.output
+    os.makedirs(dir_to_save, exist_ok=True)
+    assert os.path.isdir(dir_to_save), "Output should be a directory"
+    reports = []
+    errors_loc = os.path.join(dir_to_save, "errors.txt")
+    port = server.find_port()
+    review_config = {QualityReportAnalyzer.name: {
+        "aggregate": True,
+        }}
+
+    with tempfile.TemporaryDirectory() as tmpdirname:
+        database = args.database if args.database else os.path.join(tmpdirname, "db.sqlite3")
+        fs = args.fs if args.fs else os.path.join(tmpdirname, "models")
+        os.makedirs(fs, exist_ok=fs)
+        with AnalyzerContextManager(port=port, db=database, fs=fs,
+                                    analyzer="lookout.style.format.quality_report", init=False):
+            for repo in REPOSITORIES:
+                repo, to_commit, from_commit = repo.split()
+                report_loc = os.path.join(dir_to_save, get_repo_name(repo))
+                quality_rep_loc = report_loc + ".quality_report.md"
+                model_rep_loc = report_loc + ".model_report.md"
+                # generate or read report
+                try:
+                    if args.force or not os.path.exists(quality_rep_loc) or \
+                            not os.path.exists(model_rep_loc):
+                        # Skip this step if report was already generated
+                        quality_report, model_report = measure_quality(
+                            repo, to_commit=to_commit, from_commit=from_commit, port=port,
+                            review_config=review_config)
+                        if quality_report is not None:
+                            with open(quality_rep_loc, "w", encoding="utf-8") as f:
+                                f.write(quality_report)
+                        if model_report is not None:
+                            with open(model_rep_loc, "w", encoding="utf-8") as f:
+                                f.write(model_report)
+                    else:
+                        with open(quality_rep_loc, encoding="utf-8") as f:
+                            quality_report = f.read()
+                        with open(model_rep_loc, encoding="utf-8") as f:
+                            model_report = f.read()
+                    report = "\n".join((quality_report, model_report))
+                    reports.append((repo, report))
+                except Exception as e:
+                    with open(errors_loc, "a") as f:
```

Comment:
I feel that we have just reinvented `logging.Logger.exception()` with an additional file handler for ERROR log level here. I would love to write "TODO" here but I just cannot leave this in the release.